### ISSUE TITLE
numerical instability under different shapes in `jnp.linalg.norm` when `ord="fro"`

### ISSUE THREAD

**Issue by bicycleman15** [google/jax#10374]
I am getting this weird behaviour in `jnp.linalg.norm` where in when I set `ord="fro"`, I get different values when I pass input array `x` having shape as `[1,1]` instead of `[2, 1, 1]`. However, it is only the case for `fro`. 
For other values of `ord`, I get consistent values.

- [x] Check for duplicate issues.
- [x] Provide a complete example of how to reproduce the bug, wrapped in triple backticks like this:

```python
import jax.numpy as jnp

x = jnp.zeros([2, 1, 1])
x = x.at[0, 0, 0].set(3.4346476e-38)
x
# DeviceArray([[[3.4346476e-38]],
#
#             [[0.0000000e+00]]], dtype=float32)

y = jnp.zeros([1, 1])
y = y.at[0, 0].set(3.4346476e-38)
y
# DeviceArray([[3.4346476e-38]], dtype=float32)

print(jnp.linalg.norm(x, "fro", (-2, -1)))
print(jnp.linalg.norm(y, "fro", (-2, -1)))
# [3.4346476e-38 0.0000000e+00]
# 0.0 -------> why is this zero ? and not 3.4346476e-38 ?

print(jnp.linalg.norm(x, 1, (-2, -1)))
print(jnp.linalg.norm(y, 1, (-2, -1)))
# [3.4346476e-38 0.0000000e+00]
# 3.4346476e-38

print(jnp.linalg.norm(x, 2, (-2, -1)))
print(jnp.linalg.norm(y, 2, (-2, -1)))
# [3.4346476e-38 0.0000000e+00]
# 3.4346476e-38
```

- [x] If applicable, include full error messages/tracebacks.

**Comment by YouJiacheng** [comment]
This behavior is related to JIT(`jnp.linalg.norm` is wrapped by `jit`)
In the `[2, 1, 1]` case
```python
jnp.sqrt(jnp.sum(jnp.real(x * jnp.conj(x)), axis=axis, keepdims=keepdims))
```
is optimized to
```python
%fused_computation (param_0.1: f32[2,1,1]) -> f32[2] {
  %param_0.1 = f32[2,1,1]{2,1,0} parameter(0)
  %abs.1 = f32[2,1,1]{2,1,0} abs(f32[2,1,1]{2,1,0} %param_0.1), metadata={op_name="jit(norm)/jit(main)/sqrt"}
  ROOT %bitcast.1 = f32[2]{0} bitcast(f32[2,1,1]{2,1,0} %abs.1), metadata={op_name="jit(norm)/jit(main)/sqrt"}
}
```
`sqrt(sum(x * x))` is optimized to `abs(x)`.
While in the `[1, 1]` case it is only optimized to `sqrt(x * x)`
```
%fused_computation (param_0.2: f32[1,1]) -> f32[] {
  %param_0.2 = f32[1,1]{1,0} parameter(0)
  %multiply.0 = f32[1,1]{1,0} multiply(f32[1,1]{1,0} %param_0.2, f32[1,1]{1,0} %param_0.2), metadata={op_name="jit(norm)/jit(main)/mul"}
  %bitcast.1 = f32[] bitcast(f32[1,1]{1,0} %multiply.0), metadata={op_name="jit(norm)/jit(main)/reduce_sum[axes=(0, 1)]"}
  ROOT %sqrt.0 = f32[] sqrt(f32[] %bitcast.1), metadata={op_name="jit(norm)/jit(main)/sqrt"}
}
```

**Comment by YouJiacheng** [comment]
@hawkinsp could you please have a look?

**Comment by hawkinsp** [comment]
Closing as duplicate of #5816

**Comment by YouJiacheng** [comment]
@hawkinsp I think the point of this issue is:
For an array of size `[2, 1, 1]` XLA DOES optimize to `abs(x)` since `sum` is over axis=(-2, -1).
BUT for an array of size `[1, 1]` XLA does NOT optimize to `abs(x)`, but `sqrt(x * x)` instead.

**Comment by hawkinsp** [comment]
Well, perhaps XLA is missing an optimization here. However, you should not be relying on it happening: XLA *may* make that optimization, but it might not for any number of reasons. The real fix is to use a numerically stable formulation of `norm`.
